substrate/client/rpc-spec-v2/src/archive/archive_storage.rs

```diff
@@ -0,0 +1,155 @@
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Implementation of the `archive_storage` method.
+
+use std::{marker::PhantomData, sync::Arc};
+
+use sc_client_api::{Backend, ChildInfo, StorageKey, StorageProvider};
+use sp_api::BlockT;
+
+use crate::common::{
+	events::{
+		ArchiveStorageMethodErr, ArchiveStorageMethodOk, ArchiveStorageResult,
+		PaginatedStorageQuery, StorageQueryType,
+	},
+	storage::{is_key_queryable, IterQueryType, QueryIter, Storage},
+};
+
+/// Generates the events of the `chainHead_storage` method.
+pub struct ArchiveStorage<Client, Block, BE> {
+	/// Storage client.
+	client: Storage<Client, Block, BE>,
+	/// The maximum number of reported items by the `archive_storage` at a time.
+	storage_max_reported_items: usize,
+	/// The maximum number of queried items allowed for the `archive_storage` at a time.
+	storage_max_queried_items: usize,
+	_phantom: PhantomData<(BE, Block)>,
+}
+
+impl<Client, Block, BE> ArchiveStorage<Client, Block, BE> {
+	/// Constructs a new [`ArchiveStorage`].
+	pub fn new(
+		client: Arc<Client>,
+		storage_max_reported_items: usize,
+		storage_max_queried_items: usize,
+	) -> Self {
+		Self {
+			client: Storage::new(client),
+			storage_max_reported_items,
+			storage_max_queried_items,
+			_phantom: PhantomData,
+		}
+	}
+}
+
+impl<Client, Block, BE> ArchiveStorage<Client, Block, BE>
+where
+	Block: BlockT + 'static,
+	BE: Backend<Block> + 'static,
+	Client: StorageProvider<Block, BE> + 'static,
+{
+	/// Generate the response of the `archive_storage` method.
+	pub fn handle_query(
+		&self,
+		hash: Block::Hash,
+		mut items: Vec<PaginatedStorageQuery<StorageKey>>,
+		child_key: Option<ChildInfo>,
+	) -> ArchiveStorageResult {
+		if let Some(child_key) = child_key.as_ref() {
+			if !is_key_queryable(child_key.storage_key()) {
+				return ArchiveStorageResult::Ok(ArchiveStorageMethodOk {
+					result: Vec::new(),
+					discarded_items: 0,
+				})
+			}
+		}
+
+		let discarded_items = items.len().saturating_sub(self.storage_max_queried_items);
+		items.truncate(self.storage_max_queried_items);
+
+		let mut storage_results = Vec::with_capacity(items.len());
+		for item in items {
+			if !is_key_queryable(&item.key.0) {
+				continue
+			}
+
+			match item.query_type {
+				StorageQueryType::Value => {
+					match self.client.query_value(hash, &item.key, child_key.as_ref()) {
+						Ok(Some(value)) => storage_results.push(value),
+						Ok(None) => continue,
+						Err(error) =>
+							return ArchiveStorageResult::Err(ArchiveStorageMethodErr { error }),
+					}
+				},
+				StorageQueryType::Hash =>
+					match self.client.query_hash(hash, &item.key, child_key.as_ref()) {
+						Ok(Some(value)) => storage_results.push(value),
+						Ok(None) => continue,
+						Err(error) =>
+							return ArchiveStorageResult::Err(ArchiveStorageMethodErr { error }),
+					},
+				StorageQueryType::ClosestDescendantMerkleValue =>
```

Comment:
Should the archive API actually support these query types?
If I see it correctly, the API for `archive_unstable_storage` is a bit different than `chainHead_unstable_storage`. The list of parameters does not have the query type for each item etc. https://paritytech.github.io/json-rpc-interface-spec/api/archive_unstable_storage.html#archive_unstable_storage

However I have not researched if this is a deliberate decision in API design or if the parameters should actually be the same as for chainHead.

Reply:
That's a good question indeed, I've jumped a bit ahead of the spec with this PR and started to implement my proposal from: https://github.com/paritytech/json-rpc-interface-spec/pull/94.

We'll still need to decide if the `ClosestDescendantMerkleValue` should be included.
From the user experience perspective, we are trying to close the gap between keeping a block pinned for as long as the user reasonably wants; and the `chainHead` stop event emitted when the blocks get too old.
Since the limits of how old / how many blocks are not entirely specified in the `chainHead` spec, we wanted to offer best-effort support and switch seamlessly between the `chainHead` and `archive` classes of methods.

That is to say, we'll pause this PR for now and make an informed design decision. 
The other alternatives are:
- specify `chainHead` limits (somehow); and 
- rebuild `archive/storage` to replay the history of the chain by returning a list of blocks and potentially a list of storage changes (my current understanding of the spec discussions)

The more people we can get to have an eye on these proposals/design decisions the better the end developer experience will be :D If you have any time I'd love to get your thoughts on the Spec repo, thanks!